# JOBSHEET 14. Fungsi 2

## Tujuan
* Mahasiswa memahami konsep fungsi rekursif
* Mahasiswa mampu mengimplementasikan fungsi rekursif dalam kode program



## Alat dan Bahan
* PC/Laptop
* Browser
* Koneksi internet
* Anaconda3 + Java kernel (opsional)

## Praktikum

### Percobaan 1
Pada percobaan ini akan dilakukan pembuatan program untuk menghitung nilai faktorial dari suatu bilangan dengan menggunakan fungsi rekursif. Selain itu, akan dibuat juga fungsi untuk menghitung nilai faktorial dengan menggunakan algoritma iteratif sebagai pembandingnya

1. Buat fungsi static dengan nama **faktorialRekursif()**, dengan tipe data kembalian fungsi int dan memiliki 1 parameter dengan tipe data int berupa bilangan yang akan dihitung nilai faktorialnya

    ![Gambar 1](images/code14-1.png)

In [1]:
// Tuliskan kode program Percobaan 1 Langkah 1
static int faktorialRekursif(int n){
    if (n == 0){
        return (1);
    } else {
        return (n * faktorialRekursif(n - 1));
    }
}

2. Buat lagi fungsi static dengan nama **faktorialIteratif()**, dengan tipe data kembalian fungsi int dan memiliki 1 parameter dengan tipe data int berupa bilangan yang akan dihitung nilai faktorialnya.

    ![Gambar 2](images/code14-2.png)

In [14]:
// Tuliskan kode program Percobaan 1 Langkah 2
static int faktorialIteratif(int n){
    int faktor = 1;
    for (int i = n; i >= 1; i--){
        faktor = faktor * i;
    }
    return faktor;
}

3. Lakukan pemanggilan terhadap kedua fungsi yang telah dibuat sebelumnya, dan tampilkan hasil yang didapatkan.

    ![Gambar 3](images/code14-3.png)

In [15]:
// Tuliskan kode program Percobaan 1 Langkah 1, 2, 3
System.out.println(faktorialRekursif(5));
System.out.println(faktorialIteratif(5));

120
120


4. Jika ditelusuri, pada saat pemanggilan fungsi faktorialRekursif(5),maka proses yang terjadi dapat diilustrasikan sebagai berikut:

    ![Gambar 4](images/code14-4.png)

#### Pertanyaan
1. Apa yang dimaksud dengan fungsi rekursif?

- fungsi rekursif adalah fungsi yang memanggil dirinya sendiri sehingga fungsi akan dilakukan berulang ulang hingga kondisi base case terpenuhi.

2. Bagaimana contoh kasus penggunaan fungsi rekursif ?

- dengan membuat fungsi yang didalamnya terdapat if digunakan untuk base case dan else digunakan untuk recursion call. contoh kasus fungsi rekursif digunakan untuk menghitung perpangkatan dan faktorial.

3. Pada Percobaan1, apakah hasil yang diberikan fungsi faktorialRekursif() dan fungsi faktorialIteratif() sama? Jelaskan perbedaan alur jalannya program pada penggunaan fungsi rekursif dan fungsi iteratif!

- hasilnya sama

- untuk fungsi faktorialRekursif() operasi akan dilakukan secara bertingkat. misal faktorial nya dilakukan 5 kali. maka dari 2 kali 1 = 2, 2 dikali 3 = 6, 6 dikali 4 = 24, 24 dikali 5 = 120. jadi operasi dilakukan 1 per 1 hingga base case terpenuhi.

- untuk fungsi faktorialIteratif() menggunakan for dilakukan di 1 tingkatan saja. jadi perluangan nya terjadi 1*2*3*4*5 = 120.


### Percobaan 2
Pada percobaan ini akan dilakukan pembuatan program untuk menghitung pangkat sebuah bilangan dengan menggunakan fungsi rekursif.

1. Buat fungsi static dengan nama **hitungPangkat()**, dengan tipe data kembalian fungsi int dan memiliki 2 parameter dengan tipe data int berupa bilangan yang akan dihitung pangkatnya dan bilangan pangkatnya

    ![Gambar 5](images/code14-5.png)

2.	Deklarasikan Scanner dengan nama sc
3.	Buatlah dua buah variabel bertipe int dengan nama bilangan dan pangkat
4.	Tambahkan kode berikut ini untuk menerima input dari keyboard

    ![Gambar 6](images/code14-6.png)

5. Lakukan pemanggilan fungsi hitungPangkat yang telah dibuat sebelumnya dengan mengirimkan dua nilai parameter.

    ![Gambar 7](images/code14-7.png)

In [9]:
// Tuliskan kode program Percobaan 2 Langkah 1 - 5
static int hitungPangkat(int x, int y){
    if (y == 0){
        return (1);
    }else {
        return (x * hitungPangkat(x, y-1));
    }
}

import java.util.Scanner;
Scanner sc = new Scanner(System.in);

int bilangan, pangkat;

System.out.print("Bilangan yang dihitung: ");
bilangan = sc.nextInt();
System.out.print("Pangkat: ");
pangkat = sc.nextInt();

System.out.println(hitungPangkat(bilangan, pangkat));

Bilangan yang dihitung: 2
Pangkat: 3
8


#### Pertanyaan
1. Pada Percobaan2, terdapat pemanggilan fungsi rekursif hitungPangkat(bilangan, pangkat) pada fungsi main, kemudian dilakukan pemanggilan fungsi hitungPangkat() secara berulangkali. Jelaskan sampai kapan proses pemanggilan fungsi tersebut akan dijalankan!

Jawab

- pada setiap pemanggilan nilai y(pangkat) akan dikurang 1, apabila y(pangkat) sudah menjadi 0 maka fungsi tidak akan memanggil dirinya lagi tetapi mencetak 1. setelah itu dilakukan operasinya

### Percobaan 3
Pada percobaan ini akan dilakukan pembuatan program untuk menghitung jumlah uang nasabah yang disimpan di Bank setelah mendapatkan bunga selama beberapa tahun dengan menggunakan fungsi rekursif. 

1. Buat fungsi static dengan nama **hitungBunga()**, dengan tipe data kembalian fungsi double dan memiliki 2 parameter dengan tipe  data int berupa saldo nasabah dan lamanya menabung. Pada kasus ini dianggap bunga yang ditentukan oleh bank adalah 11% per tahun. Karena perhitungan bunga adalah bunga * saldo, sehingga untuk menghitung besarnya uang setelah ditambah bunga adalah saldo + bunga * saldo. Dalam hal ini, besarnya bunga adalah 0.11 * saldo, dan saldo dianggap 1 * saldo, sehingga 1 * saldo + 0.11 * saldo dapat diringkas menjadi 1.11 * saldo untuk perhitungan saldo setelah ditambah bunga (dalam setahun).

    ![Gambar 8](images/code14-8.png)

2.	Deklarasikan Scanner dengan nama sc
3.	Buatlah sebuah variabel bertipa double dengan nama saldoAwal dan sebuah variabel bertipe int bernama tahun
4.	Tambahkan kode berikut ini untuk menerima input dari keyboard

 ![Gambar 9](images/code14-9.png)

5. Lakukan pemanggilan fungsi hitungBunga yang telah dibuat sebelumnya dengan mengirimkan dua nilai parameter.

    ![Gambar 10](images/code14-10.png)

In [12]:
// Tuliskan kode program Percobaan 3 Langkah 1 - 5
static double hitungBunga(double saldo, int tahun){
    if(tahun == 0){
        return (saldo);
    }else {
        return(1.11 * hitungBunga(saldo, tahun-1));
    }
}

import java.util.Scanner;
Scanner sc = new Scanner(System.in);

double saldoAwal;
int tahun;

System.out.print("Jumlah saldo awal: ");
saldoAwal = sc.nextInt();
System.out.print("Lamanya Menabung (tahun): ");
tahun = sc.nextInt();

System.out.print("Jumlah uang setelah " + tahun + " tahun: ");
System.out.println(hitungBunga(saldoAwal, tahun));

Jumlah saldo awal: 300000
Lamanya Menabung (tahun): 3
Jumlah uang setelah 3 tahun: 410289.30000000016


#### Pertanyaan
1. Pada Percobaan3, sebutkan blok kode program manakah yang merupakan “base case” dan “recursion call”!

didalam fungsi tersebut, 
- if(tahun == 0){return (saldo)} adalah base case.
- else {return(1.11 * hitungBunga(saldo, tahun-1));} adalah recursion call.

## Tugas

1. Buatlah program untuk menampilkan bilangan n sampai 0 dengan menggunakan fungsi rekursif dan fungsi iteratif. (**DeretDescendingRekursif**).

In [8]:
// Tuliskan jawaban tugas nomor 1
static int tugas1a(int n){
    if(n==0){
        return (0);
    } else {
        System.out.println(n);
        return (tugas1a(n-1));
    }
}

static int tugas1b(int n){
    int hasil = n;
    for(int i = n; i>0; i--){
        System.out.println(hasil);
        hasil -= 1;
    }
    return (hasil);
}

Scanner input1 = new Scanner(System.in);
System.out.print("Masukkan Nilai: ");
int nilai = input1.nextInt();

System.out.println("Fungsi Rekursif");
System.out.println(tugas1a(nilai));

System.out.println("\nFungsi Iteratif");
System.out.println(tugas1b(nilai));

Masukkan Nilai: 5
Fungsi Rekursif
5
4
3
2
1
0

Fungsi Iteratif
5
4
3
2
1
0


2. Buatlah program yang di dalamnya terdapat fungsi rekursif untuk menghitung bilangan faktorial. Misalniya f = 8, maka akan dihasilkan 1+2+3+4+5+6+7+8 = 36 (**PenjumlahanRekursif**).

In [11]:
// Tuliskan jawaban tugas nomor 2
static int penjumlahanRekursif(int f){
    if(f == 0){
        return (0);
    }else {
        System.out.print(f + " + ");
        return(f + penjumlahanRekursif(f-1));
    }
}

Scanner input1 = new Scanner(System.in);
System.out.print("Masukkan bilangan: ");
int nilai = input1.nextInt();

int hasil = penjumlahanRekursif(nilai);

System.out.println("= " + hasil);

Masukkan bilangan: 8
8 + 7 + 6 + 5 + 4 + 3 + 2 + 1 + = 36


3.	Buat program yang di dalamnya terdapat fungsi rekursif untuk mengecek apakah suatu bilangan n merupakan bilangan prima atau bukan. n dikatakan bukan bilangan prima jika ia habis dibagi dengan bilangan kurang dari n. (**CekPrimaRekursif**).

In [12]:
// Tuliskan jawaban tugas nomor 3
static int rekursif(int a, int b){
    if(b == 1){
        return (1);
    }else if(a % b == 0){
        return (1 + rekursif(a, --b));
    }else {
        return (0 + rekursif(a, --b));
    }
}

static boolean cek(int c){
    if(c>1){
        return(rekursif(c, c)==2);
    }else{
        return false;
    }
}

int nilai;
Scanner input = new Scanner(System.in);
System.out.print("Masukkan nilai: ");
nilai = input.nextInt();

if(cek(nilai) == true){
    System.out.println(nilai + " Adalah Bilangan Prima");
}else{
    System.out.println(nilai + " Bukan Bilangan Prima");
}


Masukkan nilai: 83
83 Adalah Bilangan Prima


4.	Sepasang marmut yang baru lahir (jantan dan betina) ditempatkan pada suatu pembiakan.  Setelah dua bulan pasangan marmut tersebut melahirkan sepasang marmut kembar (jantan dan betina). Setiap pasangan marmut yang lahir juga akan melahirkan sepasang marmut juga setiap 2 bulan.  Berapa pasangan marmut yang ada pada akhir bulan ke-12? Buatlah programnya menggunakan fungsi rekursif! (**Fibonacci**).
Berikut ini adalah ilustrasinya dalam bentuk tabel.

 ![Gambar 11](images/code14-11.PNG)

In [13]:
// Tuliskan jawaban tugas nomor 4
static int Fibonaci(int n){
    if(n == 0 || n == 1){
        return (n);
    }else{
        return (Fibonaci(n-1) + Fibonaci(n-2));
    }
}

Scanner input = new Scanner(System.in);
System.out.print("Masukkan bulan: ");
int nilai = input.nextInt();

System.out.println("total pasangan marmut pada akhir bulan ke-"+ nilai +" adalah " +Fibonaci(nilai) + " Pasangan.");

Masukkan bulan: 12
total pasangan marmut pada akhir bulan ke-12 adalah 144 Pasangan.
